From: Tomaz Zaman <tomaz@mono.si>
Date: Tue, 21 Oct 2025 12:00:00 +0000
Subject: [PATCH] hwmon: ina2xx: Add INA234 support as INA226 variant

Add support for the Texas Instruments INA234 power monitor.

The INA234 is a 12-bit device with similar register layout to INA226.
The key difference is the bus voltage LSB: INA234 uses 3.125mV which
becomes 1.6mV effective LSB due to the 12-bit format (data in bits
[15:4] with bits [3:0] reserved).

Datasheet: https://www.ti.com/product/ina234

Upstream-Status: Inappropriate [embedded specific]
Signed-off-by: Tomaz Zaman <tomaz@mono.si>
---
 drivers/hwmon/ina2xx.c | 27 ++++++++++++++++++++-------
 1 file changed, 20 insertions(+), 7 deletions(-)

diff --git a/drivers/hwmon/ina2xx.c b/drivers/hwmon/ina2xx.c
index xxxxxxxx..xxxxxxxx 100644
--- a/drivers/hwmon/ina2xx.c
+++ b/drivers/hwmon/ina2xx.c
@@ -125,7 +125,7 @@
 	.writeable_reg = ina2xx_writeable_reg,
 };
 
-enum ina2xx_ids { ina219, ina226 };
+enum ina2xx_ids { ina219, ina226, ina234 };
 
 struct ina2xx_config {
 	u16 config_default;
@@ -164,6 +164,14 @@
 		.bus_voltage_lsb = 1250,
 		.power_lsb_factor = 25,
 	},
+	[ina234] = {
+		.config_default = INA226_CONFIG_DEFAULT,
+		.calibration_value = 2048,
+		.shunt_div = 400,
+		.bus_voltage_shift = 0,
+		.bus_voltage_lsb = 1600,
+		.power_lsb_factor = 25,
+	},
 };
 
 /*
@@ -633,12 +641,12 @@
 			return 0444;
 		case hwmon_in_lcrit:
 		case hwmon_in_crit:
-			if (chip == ina226)
+			if (chip == ina226 || chip == ina234)
 				return 0644;
 			break;
 		case hwmon_in_lcrit_alarm:
 		case hwmon_in_crit_alarm:
-			if (chip == ina226)
+			if (chip == ina226 || chip == ina234)
 				return 0444;
 			break;
 		default:
@@ -651,12 +659,12 @@
 			return 0444;
 		case hwmon_curr_lcrit:
 		case hwmon_curr_crit:
-			if (chip == ina226)
+			if (chip == ina226 || chip == ina234)
 				return 0644;
 			break;
 		case hwmon_curr_lcrit_alarm:
 		case hwmon_curr_crit_alarm:
-			if (chip == ina226)
+			if (chip == ina226 || chip == ina234)
 				return 0444;
 			break;
 		default:
@@ -668,11 +676,11 @@
 		case hwmon_power_input:
 			return 0444;
 		case hwmon_power_crit:
-			if (chip == ina226)
+			if (chip == ina226 || chip == ina234)
 				return 0644;
 			break;
 		case hwmon_power_crit_alarm:
-			if (chip == ina226)
+			if (chip == ina226 || chip == ina234)
 				return 0444;
 			break;
 		default:
@@ -682,7 +690,7 @@
 	case hwmon_chip:
 		switch (attr) {
 		case hwmon_chip_update_interval:
-			if (chip == ina226)
+			if (chip == ina226 || chip == ina234)
 				return 0644;
 			break;
 		default:
@@ -802,7 +810,7 @@
 	if (ret < 0)
 		return ret;
 
-	if (data->chip == ina226) {
+	if (data->chip == ina226 || data->chip == ina234) {
 		bool active_high = device_property_read_bool(dev, "ti,alert-polarity-active-high");
 
 		regmap_update_bits(regmap, INA226_MASK_ENABLE,
@@ -872,6 +880,7 @@
 	{ "ina226", ina226 },
 	{ "ina230", ina226 },
 	{ "ina231", ina226 },
+	{ "ina234", ina234 },
 	{ }
 };
 MODULE_DEVICE_TABLE(i2c, ina2xx_id);
@@ -897,6 +906,10 @@
 		.compatible = "ti,ina231",
 		.data = (void *)ina226
 	},
+	{
+		.compatible = "ti,ina234",
+		.data = (void *)ina234
+	},
 	{ },
 };
 MODULE_DEVICE_TABLE(of, ina2xx_of_match);
